src/VisualStudio/Core/Def/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs

```diff
@@ -224,6 +229,54 @@ public int RemoveAdornments()
 
                 return VSConstants.S_OK;
             }
+
+            int IVsDocOutlineProvider.GetOutline(out IntPtr phwnd, out IOleCommandTarget? ppCmdTarget)
+            {
+                var languageServiceBroker = _languageService.Package.ComponentModel.GetService<ILanguageServiceBrokerShim>();
+                var threadingContext = _languageService.Package.ComponentModel.GetService<IThreadingContext>();
+                var asyncListenerProvider = _languageService.Package.ComponentModel.GetService<IAsynchronousOperationListenerProvider>();
+                var asyncListener = asyncListenerProvider.GetListener(FeatureAttribute.DocumentOutline);
+                var editorAdaptersFactoryService = _languageService.Package.ComponentModel.GetService<IVsEditorAdaptersFactoryService>();
+
+                threadingContext.ThrowIfNotOnUIThread();
+                var documentOutlineView = new DocumentOutlineControl(languageServiceBroker, threadingContext, asyncListener, editorAdaptersFactoryService, _codeWindow);
+
+                _documentOutlineViewHost = new ElementHost
```

Comment:
this is overwriting the existing host if there is one.  is that ok?  if so, please doc.  is it expected that GetOutline can be called multiple times?  If so, doc that.  If not, check and return failure results here.

Reply:
I admit I'm still quite worried about this -- do we need a hard Contract.ThrowIfNull() to ensure we're relying on that here?